packages/next/server/base-http.ts

```diff
@@ -0,0 +1,273 @@
+  abstract statusCode: number | undefined
+  abstract statusMessage: string | undefined
+  abstract get sent(): boolean
+
+  constructor(public destination: Destination) {}
+
+  /**
+   * Sets a value for the header overwriting existing values
+   */
+  abstract setHeader(name: string, value: string): this
+
+  /**
+   * Appends value for the given header name
+   */
+  abstract appendHeader(name: string, value: string): this
+
+  /**
+   * Get all vaues for a header as an array or undefined if no value is present
+   */
+  abstract getHeaderValues(name: string): string[] | undefined
+
+  /**
+   * Get vaues for a header concatenated using `,` or undefined if no value is present
+   */
+  abstract getHeader(name: string): string | undefined
+
+  abstract body(value: string): this
+
+  abstract send(): void
+
+  // Utils implemented using the abstract methods above
+
+  redirect(destination: string, statusCode: number) {
+    this.setHeader('Location', destination)
+    this.statusCode = statusCode
+
+    if (statusCode === PERMANENT_REDIRECT_STATUS) {
+      this.setHeader('Refresh', `0;url=${destination}`)
+    }
+    return this
+  }
+}
+
+class NodeNextResponse extends BaseNextResponse<Writable> {
+  private textBody: string | undefined = undefined
+
+  constructor(public res: ServerResponse) {
+    super(res)
+  }
+
+  get sent() {
+    return this.res.finished || this.res.headersSent
+  }
+
+  get statusCode() {
+    return this.res.statusCode
+  }
+
+  set statusCode(value: number) {
+    this.res.statusCode = value
+  }
+
+  get statusMessage() {
+    return this.res.statusMessage
+  }
+
+  set statusMessage(value: string) {
+    this.res.statusMessage = value
+  }
+
+  setHeader(name: string, value: string): this {
+    this.res.setHeader(name, value)
+    return this
+  }
+
+  getHeaderValues(name: string): string[] | undefined {
+    const values = this.res.getHeader(name)
+
+    if (values === undefined) return undefined
+
+    return (Array.isArray(values) ? values : [values]).map((value) =>
+      value.toString()
+    )
+  }
+
+  getHeader(name: string): string | undefined {
+    const values = this.getHeaderValues(name)
+    return Array.isArray(values) ? values.join(',') : undefined
+  }
+
+  appendHeader(name: string, value: string): this {
+    const currentValues = this.getHeaderValues(name) ?? []
+
+    if (!currentValues.includes(value)) {
+      this.res.setHeader(name, [...currentValues, value])
+    }
+
+    return this
+  }
+
+  body(value: string) {
+    this.textBody = value
+    return this
+  }
+
+  send() {
+    this.res.end(this.textBody)
+  }
+}
+
+class WebNextResponse extends BaseNextResponse<WritableStream> {
+  private headers = new Headers()
+  private textBody: string | undefined = undefined
+  private _sent = false
+
+  private sendPromise = new Promise<void>((resolve) => {
+    this.sendResolve = resolve
+  })
+  private sendResolve?: () => void
+  private response = this.sendPromise.then(() => {
```

Comment:
This will behave same as Node response. This promise will resolve when `send` is called with a Response object that has the headers/status at that moment, but the body will still be able to be streamed.

To match 100%, we could add a call to `send` within the transform stream when the first write happens, but since we'll only be using streams in `render.tsx` we can just call `send` there manually for simplicity